### ISSUE TITLE
lax.conv_transpose has different convention than TF, document it and make switching easy?

### ISSUE THREAD

**Issue by mattjj** [google/jax#14337]
See screenshot:
<img width="671" alt="image" src="https://user-images.githubusercontent.com/1458824/217355716-0cec239a-afad-4fe8-8442-995240a3fe95.png">

Though the example uses Haiku, Haiku is just calling through to `lax.conv_transpose`.

The difference is in reversing the spatial axes of the kernel.

We can do at least two things to make this less surprising and more easily fixable:
1. document the difference, and
2. add an argument to `lax.conv_transpose` to instead use the TF convention.

**Comment by MaanasArora** [comment]
Hi @mattjj,

I was looking at this issue and found this documentation on line 293 of [jax/_src/lax/convolution.py](https://github.com/google/jax/blob/main/jax/_src/lax/convolution.py):

>     transpose_kernel: if True flips spatial axes and swaps the input/output
>       channel axes of the kernel. This makes the output of this function identical
>       to the gradient-derived functions like keras.layers.Conv2DTranspose
>       applied to the same kernel. For typical use in neural nets this is completely
>       pointless and just makes input/output channel specification confusing.

Any chance 2 has been implemented already? I agree the documentation could be more conspicuous.

Thank you!

**Comment by NaokiDohi** [comment]
@mattjj Hi, I find it as good first issue. I want to contribute this repo for the my first github contribution. Do u have something I can help?

**Comment by alhridoy** [comment]
Hi @mattjj I have taken a look at this issue regarding the difference in reversing the spatial axes of the kernel between Haiku's call and `lax.conv_transpose`. I understand the potential sources of confusion and believe that the proposed solutions, both documenting the difference and adding an argument for the TF convention, are valuable.

I'm eager to contribute and work on this issue.

**Comment by SinghSyntax001** [comment]
Clarification on tf_style Argument in conv_transpose

Hi everyone,

I recently submitted a PR to add the tf_style: bool = False argument to the conv_transpose function, but it was closed because the new argument was unused. I’d like to clarify my intent and see if this feature is necessary.

The goal of adding tf_style was to ensure compatibility with TensorFlow-style transposed convolutions. Specifically, I wanted to provide an option that aligns the behavior with TensorFlow’s implementation when needed. However, I realize now that I didn’t incorporate it fully into the function logic.

Before I resubmit a PR, I’d like to ask:

Would this feature be useful? Are there cases where users might need conv_transpose to behave like TensorFlow’s version?
How should it be implemented? If maintaining TensorFlow compatibility is desirable, should this flag modify the function behavior, or is there already another approach in place?
Looking forward to your feedback! If this feature isn’t needed, I’ll move on to other contributions.

Thanks!

**Comment by Rakshars** [comment]
Hey what is the status of this issue?

**Comment by tarajyothi** [comment]
Hi @mattjj,

I’m a student preparing for GSOC 2026 and would like to contribute to JAX.
I’d like to work on this issue from a documentation perspective.

My plan is to:
- Make the JAX vs TensorFlow kernel convention difference more explicit in the docs/docstring
- Add a small illustrative example to reduce user confusion

Please let me know if this approach sounds good. Thanks!

**Comment by rosspeili** [comment]
Is this closed?

**Comment by Mudassir55** [comment]
Hi, I am a beginner contributor and I would like to work on the documentation part of this issue.

I noticed that `lax.conv_transpose` already has a `transpose_kernel` argument, but the TensorFlow/Keras compatibility behavior could still be made clearer with a small example.

Would a documentation PR that adds a clearer explanation and example for `transpose_kernel=True` be useful?

**Comment by wasup-yash** [comment]
I think this issue is already addressed, right? So we should close it

**Comment by wasim-builds** [comment]
Hi! I'm a beginner looking to
  make a contribution. Could I please work on the documentation part of this
  issue to clarify the difference in spatial axes compared to TensorFlow?

**Comment by kodlan** [comment]
It seems like both things the issue asked for appear to be done:

1. **An argument to switch to the TF convention already exists**: `transpose_kernel=True` flips the kernel's spatial axes and swaps the input/output channel axes, which is exactly the TF/Keras `Conv2DTranspose` convention. 

2. **The difference is now documented**: commit 7a57a4cd5 (Nov 2025) added an explicit note to the `lax.conv_transpose` docstring:
   > TensorFlow/Keras Compatibility: By default, JAX does NOT reverse the kernel's spatial dimensions. This differs from TensorFlow's "Conv2DTranspose" and similar frameworks, which flip spatial axes and swap input/output channels.
   >
   > To match TensorFlow/Keras behavior, set `transpose_kernel=True`.

Seems like this issue can be closed.

The only leftover I noticed is that the `transpose_kernel` parameter docstring still says using it is "completely pointless", which now contradicts the Notes section above it that recommends the flag for TF compatibility — I'll create a small PR for this.

**Comment by LeonxLJX** [comment]
I'd like to take this one.

From the report, the useful first step is to pin down the exact code path involved before touching anything — these tend to have a small true cause behind a broad-looking symptom, and guessing at it usually produces a patch that masks the behavior instead of fixing it.

Plan: trace the path, reproduce it, then make a focused change with a test if this area has coverage. I'll keep the diff small so it's easy to review.

If there's an existing convention in this repo for how this kind of thing is handled, I'd rather follow it than invent my own — happy to be pointed at one.

(claiming via @LeonxLJX)
